learning_wavelets/evaluate/utils.py

```diff
@@ -0,0 +1,72 @@
+import numpy as np
+import tensorflow as tf
+
+
+from runstats import Statistics
+from skimage.measure import compare_psnr, compare_ssim
+
+
+
+tf.random.set_seed(1)
+
+def mse(gt, pred):
+    """ Compute Mean Squared Error (MSE) """
+    return np.mean((gt - pred) ** 2)
+
+
+def nmse(gt, pred):
+    """ Compute Normalized Mean Squared Error (NMSE) """
+    return np.linalg.norm(gt - pred) ** 2 / np.linalg.norm(gt) ** 2
+
+
+def psnr(gt, pred):
+    """ Compute Peak Signal to Noise Ratio metric (PSNR) """
+    return compare_psnr(gt, pred, data_range=gt.max() - gt.min())
+
+
+def ssim(gt, pred):
+    """ Compute Structural Similarity Index Metric (SSIM). """
+    return compare_ssim(
+        gt.transpose(1, 2, 0), pred.transpose(1, 2, 0), multichannel=True, data_range=gt.max() - gt.min()
+    )
+
+
+METRIC_FUNCS = dict(
+    LOSS=mse,
+    PSNR=psnr,
+    SSIM=ssim,
+)
+
+
+class Metrics:
+    """
+    Maintains running statistics for a given collection of metrics.
+    """
+
+    def __init__(self, metric_funcs):
+        self.metrics = {
+            metric: Statistics() for metric in metric_funcs
+        }
+        self.metric_funcs = metric_funcs
+
+    def push(self, target, recons):
+        for metric, func in self.metric_funcs.items():
+            self.metrics[metric].push(func(target, recons))
+
+    def means(self):
+        return {
+            metric: stat.mean() for metric, stat in self.metrics.items()
+        }
+
+    def stddevs(self):
+        return {
+            metric: stat.stddev() for metric, stat in self.metrics.items()
+        }
+
+    def __repr__(self):
+        means = self.means()
+        stddevs = self.stddevs()
+        metric_names = sorted(list(means))
+        return ' '.join(
+            f'{name} = {means[name]:.4g} +/- {stddevs[name]:.4g}' for name in metric_names
+        )
```

Comment:
I think your editor has a problem with trailing whitespaces etc.

Try to configure it to remove trailing whitespaces and to always have a line return at the end of the file.

I think more generally you should consider moving away from spyder and use something more modern like pycharm, vscode or atom this way you will be more efficient